### PR TITLE
Emmett bot

### PR DESCRIPTION
**Pull request by Emmett-Byrne**

### CODE REVIEW

**Comment by PrzemekTomczyk** [comment]
you're gonna have to wait until Cathal's screens are merged in

File: ARGO/include/AiSystem.h [hunk 1]
```diff
@@ -20,5 +23,25 @@ class AiSystem : public BaseSystem
...
+	void setEnemyData(glm::vec2 t_botPosition);
+	void setClosestLeaderData(glm::vec2 t_botPosition);
+	void setClosestPickupData(glm::vec2 t_botPosition);
+	void setGoalData(glm::vec2 t_botPosition);
+
+	Entity(&m_players)[Utilities::S_MAX_PLAYERS];
+	std::vector<Entity>& m_enemies;
+	EventManager& m_eventManager;
+
+	WeightedSelector m_behaviourTree;
```
Comment (CathalRedmond) [inline ARGO/include/AiSystem.h:40]: Is this a behaviour tree though?

File: ARGO/include/BehaviourTree.h [hunk 2]
```diff
@@ -0,0 +1,172 @@
+#pragma once
+#include <Entity.h>
+
+////----------------------------This is the data we will need for our behaviour to function.----------------------------
+const int BOT_CAN_SEE_ENEMY_DISTANCE = Utilities::TILE_SIZE * 8;
+const int BOT_DISTANCE_TO_CARE_ABOUT_GOAL_SCALER = Utilities::TILE_SIZE * 2;
+const int BOT_DEFAUL_MOVE_TOWARD_GOAL_WEIGHT = 5;
```
Comment (CathalRedmond) [inline ARGO/include/BehaviourTree.h]: 'DEFAUL' ?

Comment (CathalRedmond) [reply ARGO/include/BehaviourTree.h]: Also do you need to add bot to the start of all of these?

File: ARGO/include/BehaviourTree.h [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+{
+public:
+	virtual bool run(Entity& t_entity) override
+	{
+		WeightedNode* bestNode = nullptr;
+		float bestWeight = -1.0f;
+		for (WeightedNode* child : getChildren()) 
+		{
+			float weight = child->getWeight();
+			if (weight > bestWeight)
```
Comment (CathalRedmond) [inline ARGO/include/BehaviourTree.h:59]: why not just have the 'child->getWeight()' here instead of 'weight'

File: ARGO/include/BehaviourTree.h [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+{
+	Entity* entity;
+	float distance;
+};
+struct GoalData
+{
+	Entity* entity;
+	float distance;
+};
+
```
Comment (CathalRedmond) [inline ARGO/include/BehaviourTree.h:31]: These 3 structs share the same data they can probably be made into one struct

File: ARGO/include/BehaviourTree.h [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+		if (forceComp)
+		{
+			forceComp->addForce(glm::normalize(pos - enemyPos));
+			return true;
+		}
+		return false;
+	}
+	virtual float getWeight() override
+	{
+		return (1 - m_data->distance / (BOT_CAN_SEE_ENEMY_DISTANCE * BOT_CAN_SEE_ENEMY_DISTANCE)) * 10;
```
Comment (CathalRedmond) [inline ARGO/include/BehaviourTree.h]: magic numbers 1 and 10

File: ARGO/include/BehaviourTree.h [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+	{
+		return true;
+	}
+	virtual float getWeight() override
+	{
+		return m_data->nearbyEnemies + 1.0f;
+	}
+};
+
+class GetAmmoBehaviour : public WeightedNode { //TODO
```
Comment (CathalRedmond) [inline ARGO/include/BehaviourTree.h]: delete comment

Comment (CathalRedmond) [reply ARGO/include/BehaviourTree.h]: same on line 132

File: ARGO/include/BehaviourTree.h [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+		if (forceComp && goalTransComp->getPos() != pos)
+		{
+			forceComp->addForce(glm::normalize(goalTransComp->getPos() - pos));
+			return true;
+		}
+		return false;
+	}
+	virtual float getWeight() override
+	{
+		if (m_data->distance == 0)
```
Comment (CathalRedmond) [inline ARGO/include/BehaviourTree.h]: tertiary operator could be used here

File: ARGO/src/AiSystem.cpp [hunk 8]
```diff
@@ -58,6 +73,89 @@ void AiSystem::rangedAI(TransformComponent* t_posComp, AiComponent* t_aiComponen
...
+					m_botLeaderData.distance = newDistance;
+				}
+			}
+		}
+	}
+}
+
+void AiSystem::setClosestPickupData(glm::vec2 t_botPosition)
+{
+	//TODO
```
Comment (CathalRedmond) [inline ARGO/src/AiSystem.cpp]: is commented needed

File: ARGO/src/Game.cpp [hunk 9]
```diff
@@ -62,11 +63,41 @@ Game::Game() :
 		}
 
 		m_levelManager.setupLevel();
```
Comment (CathalRedmond) [inline ARGO/src/Game.cpp]: Too many magic numbers

File: ARGO/src/Game.cpp [hunk 10]
```diff
@@ -330,11 +363,20 @@ void Game::createPlayer(Entity& t_player)
 		std::pair<ButtonType, Command*>(ButtonType::RightTrigger, new FireBulletCommand()),
 		std::pair<ButtonType,Command*>(ButtonType::Back, new CloseWindowCommand()) };
 
+	InputComponent* impComp = new InputComponent(buttonPressMap, buttonPressMap);
+	if (impComp->getController().getSDLController())
+	{
+		// passing two of the same object as at this moment the commands for button press is the same for button held
```
Comment (CathalRedmond) [inline ARGO/src/Game.cpp]: remove this comment

File: ARGO/src/Game.cpp [hunk 11]
```diff
@@ -330,11 +363,20 @@ void Game::createPlayer(Entity& t_player)
 		std::pair<ButtonType, Command*>(ButtonType::RightTrigger, new FireBulletCommand()),
 		std::pair<ButtonType,Command*>(ButtonType::Back, new CloseWindowCommand()) };
 
+	InputComponent* impComp = new InputComponent(buttonPressMap, buttonPressMap);
```
Comment (CathalRedmond) [inline ARGO/src/Game.cpp]: imp?

**Review by CathalRedmond — CHANGES_REQUESTED** [review]

File: ARGO/include/BehaviourTree.h (hunk 5 above)
Comment (PrzemekTomczyk) [reply ARGO/include/BehaviourTree.h]: At least have a comment explaining wtf are you doing here @NerakaiStudent 👍

**Review by PrzemekTomczyk — CHANGES_REQUESTED** [review]

File: ARGO/src/GameScreen.cpp [hunk 12]
```diff
@@ -266,10 +304,10 @@ void GameScreen::initialise(ButtonCommandMap t_controllerButtonMaps[Utilities::N
 
 void GameScreen::removeDeadEnemies()
 {
-	std::vector<Entity>::iterator iter = std::remove_if(m_entities.begin(), m_entities.end(), cleanUpEnemies);
+	/*std::vector<Entity>::iterator iter = std::remove_if(m_entities.begin(), m_entities.end(), cleanUpEnemies);
```
Comment (CathalRedmond) [inline ARGO/src/GameScreen.cpp]: is commented code needed?

**Review by CathalRedmond — CHANGES_REQUESTED** [review]

**Pull request merged**